### ISSUE TITLE
question about using JWT cookie storage

### ISSUE THREAD

**Issue by MassimoCappellano** [dwyl/hapi-auth-jwt2-cookie-example#2]
Running the example on my computer, from chrome doing two GET request (login and restricted resource):
1) FIRST:
http://localhost:8000/auth
in the server response, in the header set cookie: token=...................................... as expected
2) THEN:
http://localhost:8000/restricted
{"statusCode":401,"error":"Unauthorized","message":"Missing authentication"}

I expected that the cookie setted by the auth request(1) was set by the browser to restricted and so It should be authenticated. Using chrome console seems that the cookie of the first response
is not used in the second request.
It's correct?

Can someone explain if I miss something?

For example from this article reading: https://stormpath.com/blog/where-to-store-your-jwts-cookies-vs-html5-web-storage when explain 'JWT Cookie Storage'

**Comment by MassimoCappellano** [comment]
Because the application don't use https. I've set `isSecure: false` in `cookie_options`: in this way accessing `/restricted`, after `/auth`,  the cookie is decoded but it seams that `return callback(rediserror, true);` don't work correctly in  `function validate`. Response is:` {"statusCode":401,"error":"Unauthorized","message":"Missing authentication"}`

**Comment by nelsonic** [comment]
@MassimoCappellano _really_ good question! would you mind pushing your code up to GitHub on a branch so we can help with debugging/understanding?
(_I've invited you to join @DWYL and you have contributor status on this repository so you can create a branch right here which we can collaborate on!_) 😉

**Comment by MassimoCappellano** [comment]
Thanks @nelsonic , I've done a branch `issue_jwt_cookie_storange` with the little modifications. I hope I've done correctly, I'm not very expert.

**Comment by nelsonic** [comment]
@MassimoCappellano please open a Pull Request with your branch. often it can be _insightful_ to let the CI _confirm_ that all _existing_ tests are working on a branch and lets the people you are collaborating with see what the changes are in comparison to `master` and where we need to focus our attention. thanks! 😉

**Comment by nelsonic** [comment]
@MassimoCappellano I've added a line to the `cookie_options` in `server.js` (_see the PR_)
Now when we view the `/auth` route it sets the cookie for the domain.
![hapi-auth-jwt2-showing-auth-route](https://cloud.githubusercontent.com/assets/194400/20802422/e8468b8a-b7e3-11e6-9f6a-05989d128131.png)

And when we visit `/restricted` which _requires_ as JWT we see:
![hapi-auth-jwt2-showing-restricted-route](https://cloud.githubusercontent.com/assets/194400/20802426/eb46635a-b7e3-11e6-9cf9-ff4d09454a87.png)

Please _confirm_ you see this by pulling down the branch and running `npm install && npm start` 
thanks! 👍

**Comment by MassimoCappellano** [comment]
Hi @nelsonic thanks for help,

I've done the pull from the branch with your configuration improvement. I've the problems with cookie but  some more information. I work with windows 10 and chrome, and here some tests.

Fist doing request `/auth` and then `/restricted`, but seems not to work:

![auth_request1](https://cloud.githubusercontent.com/assets/430742/20811575/85a1764a-b80e-11e6-9179-2e9821f189d0.PNG)

![restricted_request](https://cloud.githubusercontent.com/assets/430742/20811533/60c8fc12-b80e-11e6-95c0-a9ddb1e47f17.PNG)

The strange thing is than on the console the application read the cookie correctly and find on redis the uuid.

![last_chrome](https://cloud.githubusercontent.com/assets/430742/20811767/38226694-b80f-11e6-9b9b-1e0777b3d902.PNG)

Maybe I told me depends from some extension on my browser (because I don't know what to think).

I've installalled Firefox, first request `/auth`

![auth_request_firefox](https://cloud.githubusercontent.com/assets/430742/20811901/d54ca68c-b80f-11e6-9b6e-ed9b0bc4b0fd.PNG)

![auth_request_firefox_console](https://cloud.githubusercontent.com/assets/430742/20811958/07b4271c-b810-11e6-8db1-53d80556357f.PNG)

then request `/restricted`

![restricted_request_first_firefox](https://cloud.githubusercontent.com/assets/430742/20812014/4b024fbc-b810-11e6-8e6d-da95833bc4f1.PNG)

![restricted_request_first_firefox_console](https://cloud.githubusercontent.com/assets/430742/20812027/5b92a5fc-b810-11e6-9285-cd200a8a00b4.PNG)

Now seems OK but first in the console web of  firefox, you can see that in the response header there is `Set-Cookie: token=;` In yellow above; _maybe this is the problem_.

If I repeat the` /restricted` request, I've again the problem.

![restricted_request_second_firefox](https://cloud.githubusercontent.com/assets/430742/20812284/48ef8752-b811-11e6-93dc-89ba622065c2.PNG)

and in the web console:

![restricted_request_second_firefox_webconsole](https://cloud.githubusercontent.com/assets/430742/20812297/59124ea8-b811-11e6-9f05-e8be8491812c.PNG)

Maybe in some way the cookie is like deleted after the first `/restricted` request.

**Comment by nelsonic** [comment]
@MassimoCappellano that's very strange... :-(

**Comment by MassimoCappellano** [comment]
@nelsonic yes I dont't know. Only a question, even if you resubmit `/restricted` url, you have  in response body `{"statusCode":401,"error":"Unauthorized","message":"Missing authentication"}` ?

TO REMENBER with firefox, the sequence of actions and results:

GET request `/auth` ---> response body `{"text":"Check Browser Cookie or Auth Header for your Token (JWT)"}`

then first GET `/restricted` --> response body `{"text":"You used a Token!"}`

then second GET `/restricted` --> response body `{"statusCode":401,"error":"Unauthorized","message":"Missing authentication"}`

Thanks for your work.

P.S: For now I'm starting using `hapi-auth-cookie` for the web and  hapi-auth-jwt2 for json interface.

**Comment by nelsonic** [comment]
@MassimoCappellano I have pushed up another change on the branch.
Please `git pull && npm start` and try again. 
You will not _need_ to use `hapi-auth-cookie` once you are confident that this is working for you.

**Comment by MassimoCappellano** [comment]
Thanks @nelsonic , now is perfect! There is no more the problem with the cookie when I call '/restricted'. The trick was not to re-set JWT inside  '/restricted' . Thank you very much. 

I don't know if I've to merge pull request'?

**Comment by nelsonic** [comment]
@MassimoCappellano we will get @iteles to do a review/merge of the code we've collaborated on. 
thanks _again_ (_very much_) for asking the question and helping us improve this example!! 🎉
